roslaunch: ignores ROS_NAMESPACE for parameters

Migrated from https://code.ros.org/trac/ros/ticket/2380
Attachment available in code.ros.org?: No

---

It makes working with multiple robots under one master less convenient.

I tried to clone the roslaunch_ROS_NAMESPACE and to build it using catkin_make.
I got the following cmake error:

```
$ catkin_make --source ros_comm
Base path: /shared/home/meyer/ros/hydro
Source space: /shared/home/meyer/ros/hydro/ros_comm
Build space: /shared/home/meyer/ros/hydro/build
Devel space: /shared/home/meyer/ros/hydro/devel
Install space: /shared/home/meyer/ros/hydro/install
Creating symlink "/shared/home/meyer/ros/hydro/ros_comm/CMakeLists.txt" pointing to "/opt/ros/hydro/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /shared/home/meyer/ros/hydro/ros_comm -DCATKIN_DEVEL_PREFIX=/shared/home/meyer/ros/hydro/devel -DCMAKE_INSTALL_PREFIX=/shared/home/meyer/ros/hydro/install" in "/shared/home/meyer/ros/hydro/build"
####
-- The C compiler identification is GNU
-- The CXX compiler identification is GNU
-- Check for working C compiler: /usr/lib/ccache/gcc
-- Check for working C compiler: /usr/lib/ccache/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /shared/home/meyer/ros/hydro/devel
-- Using CMAKE_PREFIX_PATH: /home/meyer/ros/hydro/gazebo/devel;/opt/ros/hydro;/opt/hector/hydro/devel;/opt/hector/hydro/devel
-- This workspace overlays: /home/meyer/ros/hydro/gazebo/devel;/opt/ros/hydro
-- Found PythonInterp: /usr/bin/python (found version "2.7.3")
-- Looking for include files CMAKE_HAVE_PTHREAD_H
-- Looking for include files CMAKE_HAVE_PTHREAD_H - found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE 
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- Using CATKIN_TEST_RESULTS_DIR: /shared/home/meyer/ros/hydro/build/test_results
-- catkin 0.5.67
-- BUILD_SHARED_LIBS is on
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- ~~  traversing 30 packages in topological order:
-- ~~  - ros_comm (metapackage)
-- ~~  - rosconsole
-- ~~  - rosgraph
-- ~~  - rosgraph_msgs
-- ~~  - roslaunch
-- ~~  - rosmaster
-- ~~  - rosmsg
-- ~~  - rosparam
-- ~~  - rospy
-- ~~  - rosservice
-- ~~  - rostest
-- ~~  - std_srvs
-- ~~  - test_roslib_comm
-- ~~  - xmlrpcpp
-- ~~  - roscpp
-- ~~  - rosout
-- ~~  - message_filters
-- ~~  - rosnode
-- ~~  - rostopic
-- ~~  - roswtf
-- ~~  - test_roscpp
-- ~~  - test_rosgraph
-- ~~  - test_roslaunch
-- ~~  - test_rosmaster
-- ~~  - test_rosparam
-- ~~  - test_rospy
-- ~~  - test_rosservice
-- ~~  - topic_tools
-- ~~  - rosbag
-- ~~  - test_rosbag
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- +++ processing catkin metapackage: 'ros_comm'
-- ==> add_subdirectory(ros_comm)
-- +++ processing catkin package: 'rosconsole'
-- ==> add_subdirectory(tools/rosconsole)
-- Boost version: 1.46.1
-- Found the following Boost libraries:
--   regex
--   thread
-- +++ processing catkin package: 'rosgraph'
-- ==> add_subdirectory(tools/rosgraph)
-- +++ processing catkin package: 'rosgraph_msgs'
-- ==> add_subdirectory(messages/rosgraph_msgs)
-- rosgraph_msgs: 2 messages, 0 services
-- +++ processing catkin package: 'roslaunch'
-- ==> add_subdirectory(tools/roslaunch)
-- +++ processing catkin package: 'rosmaster'
-- ==> add_subdirectory(tools/rosmaster)
-- +++ processing catkin package: 'rosmsg'
-- ==> add_subdirectory(tools/rosmsg)
-- +++ processing catkin package: 'rosparam'
-- ==> add_subdirectory(tools/rosparam)
-- +++ processing catkin package: 'rospy'
-- ==> add_subdirectory(clients/rospy)
-- +++ processing catkin package: 'rosservice'
-- ==> add_subdirectory(tools/rosservice)
-- +++ processing catkin package: 'rostest'
-- ==> add_subdirectory(tools/rostest)
-- +++ processing catkin package: 'std_srvs'
-- ==> add_subdirectory(messages/std_srvs)
-- std_srvs: 0 messages, 1 services
-- +++ processing catkin package: 'test_roslib_comm'
-- ==> add_subdirectory(test/test_roslib_comm)
-- +++ processing catkin package: 'xmlrpcpp'
-- ==> add_subdirectory(utilities/xmlrpcpp)
-- +++ processing catkin package: 'roscpp'
-- ==> add_subdirectory(clients/roscpp)
-- Boost version: 1.46.1
-- Found the following Boost libraries:
--   signals
--   filesystem
--   system
-- roscpp: 1 messages, 3 services
-- Looking for include files HAVE_IFADDRS_H
-- Looking for include files HAVE_IFADDRS_H - found
-- Looking for trunc
-- Looking for trunc - not found
-- +++ processing catkin package: 'rosout'
-- ==> add_subdirectory(tools/rosout)
-- +++ processing catkin package: 'message_filters'
-- ==> add_subdirectory(utilities/message_filters)
-- +++ processing catkin package: 'rosnode'
-- ==> add_subdirectory(tools/rosnode)
CMake Error at /shared/home/meyer/ros/hydro/devel/share/rostest/cmake/rostest-extras.cmake:3 (_generate_function_if_testing_is_disabled):
  Unknown CMake command "_generate_function_if_testing_is_disabled".
Call Stack (most recent call first):
  /shared/home/meyer/ros/hydro/devel/share/rostest/cmake/rostestConfig.cmake:173 (include)
  /opt/ros/hydro/share/catkin/cmake/catkinConfig.cmake:72 (find_package)
  tools/rosnode/CMakeLists.txt:3 (find_package)


-- Configuring incomplete, errors occurred!
Invoking "cmake" failed
```

--- a/tools/roslaunch/resources/roscore.xml
+++ b/tools/roslaunch/resources/roscore.xml
@@ -5,7 +5,9 @@
   http://ros.org/wiki/roscore
   -->
 <launch>
-  <param name="rosversion" command="rosversion roslaunch" />
-  <param name="rosdistro" command="rosversion -d" />
-  <node pkg="rosout" type="rosout" name="rosout" respawn="true"/>
+  <group ns="/">
+    <param name="rosversion" command="rosversion roslaunch" />
+    <param name="rosdistro" command="rosversion -d" />
+    <node pkg="rosout" type="rosout" name="rosout" respawn="true"/>
+  </group>
 </launch>

--- a/tools/roslaunch/src/roslaunch/core.py
+++ b/tools/roslaunch/src/roslaunch/core.py
@@ -196,6 +196,8 @@ def setup_env(node, machine, master_uri, env=None):
 
     # add node-specific env args last as they have highest precedence
     if node:
+        if rosgraph.ROS_NAMESPACE in d:
+            del d[rosgraph.ROS_NAMESPACE]
         ns = node.namespace 
         if ns[-1] == '/':
             ns = ns[:-1]

--- a/tools/roslaunch/src/roslaunch/xmlloader.py
+++ b/tools/roslaunch/src/roslaunch/xmlloader.py
@@ -46,7 +46,7 @@ import logging
 from xml.dom.minidom import parse, parseString
 from xml.dom import Node as DomNode #avoid aliasing
 
-from rosgraph.names import make_global_ns, ns_join, is_private, is_legal_name
+from rosgraph.names import make_global_ns, ns_join, is_private, is_legal_name, get_ros_namespace
 
 from .core import Param, Node, Test, Machine, RLException
 from . import loader
@@ -662,7 +662,7 @@ class XmlLoader(loader.Loader):
             argv = sys.argv
 
         self._launch_tag(launch, ros_config, filename)
-        self.root_context = loader.LoaderContext('', filename)
+        self.root_context = loader.LoaderContext(get_ros_namespace(), filename)
         loader.load_sysargs_into_context(self.root_context, argv)
 
         if len(launch.getElementsByTagName('master')) > 0:

